lineup sol & image overhaul (#2582)
cpinitiative/usaco-guide: solutions/usaco-965.mdx

		}
		out.close();
	}
}
```

</JavaSection>
<PySection>

```py
COWS = sorted([
	'Bessie',
	'Buttercup',
	'Belinda',
	'Beatrice',
	'Bella',
	'Blue',
	'Betsy',
	'Sue'
])

cow_inds = {c: i for i, c in enumerate(COWS)}

neighbors = [[] for _ in range(len(COWS))]
with open('lineup.in') as read:
	for _ in range(int(read.readline())):
		words = read.readline().strip().split()

		# Convert the names to their index in the list
		cow1 = cow_inds[words[0]]
		cow2 = cow_inds[words[-1]]
		neighbors[cow1].append(cow2)
		neighbors[cow2].append(cow1)

order = []
added = [False for _ in range(len(COWS))]
for c in range(len(COWS)):
	"""
	Check that:
	1. This cow hasn't already been added yet.
	2. This cow could be the possible start of a chain.
	"""
	if not added[c] and len(neighbors[c]) <= 1:
		added[c] = True
		order.append(c)

		# If the chain length > 1, we keep on going
		if len(neighbors[c]) == 1:
			prev = c
			at = neighbors[c][0]
			while len(neighbors[at]) == 2:
				added[at] = True
				order.append(at)
				a, b = neighbors[at]
				at, prev = b if a == prev else a, at

			# Add the final element
			added[at] = True
			order.append(at)

with open('lineup.out', 'w') as out:
	for c in order:

> 		print(COWS[c], file=out)

Author: wdym?

id: usaco-965
source: USACO Bronze 2019 December
title: Livestock Lineup
author: Benjamin Qi, Kevin Sheng
---

[Official Analysis (C++)](http://www.usaco.org/current/data/sol_lineup_bronze_dec19.html)

## $\mathcal{O}(N)$ Solution With Graphs

> While the solutions provided in the editorial work just fine for the given bounds,

Reviewer: ```suggestion
While the solutions provided in the editorial work for the given bounds,
```

> there *is* a solution that works in just $\mathcal{O}(N)$!

Reviewer: ```suggestion
there *is* a solution that works in just $\mathcal{O}(N)$, involving graphs!
```

Notice that since the input is guaranteed to be valid, we're always going to end

> up with "chains" of cows that we can arrange as we please. Using the sample given

Reviewer: ```suggestion
up with multiple "chains" of cows that we can arrange. Using the sample given
```

Author: chains is alr plural

> in the problem, we'd get a "chain" representation like this:

Reviewer: ```suggestion
in the problem, we'd get a "chain" representation as follows
```

Author: not needed

![](usaco-965/chains.png)

Note that cows that are not part of any chain can be considered their own chains

> of length 1 for implementation purposes.

Reviewer: not needed

Author: better explicit than implicit

Reviewer: Be more consise.

Author: > better explicit than implicit

> With this representation in mind, we can iterate through the cows in

Reviewer: ```suggestion
With this graph representation in mind, we can iterate through the cows in
```

Author: why lol

Reviewer: be more formal

Author: how does this make it more formal?

> lexicographical order. When we come upon a cow that could be a possible

Reviewer: ```suggestion
lexicographical order. When we visit a cow that could be a possible
```

> start of a chain (AKA a cow that only has one required neighbor),

Reviewer: ```suggestion
root of a chain (a cow that only has one neighbor),
```

Author: why

Reviewer: too informal

Reviewer: don't use AKA

Author: i'll wait for third opinion

we go through its neighbors (adding cows as we go along) until we hit an end.

## Implementation

**Time Complexity:** $\mathcal{O}(N)$

<LanguageSection>
<CPPSection>

```cpp
#include <fstream>
#include <vector>
#include <string>
#include <algorithm>

> 

Reviewer: please use `using namespace std;`

Author: no

using std::endl;
using std::string;
using std::vector;

//CodeSnip{Cow Names}
// Lambda expression - see /general/lambda-funcs
const vector<string> COWS = []() {
	vector<string> tmp{
		"Bessie",
		"Buttercup",
		"Belinda",
		"Beatrice",
		"Bella",
		"Blue",
		"Betsy",
		"Sue"
	};
	// sort names lexicographically
	std::sort(std::begin(tmp), std::end(tmp));
	return tmp;
}();
//EndCodeSnip

template <class T>
int index(const vector<T>& vec, const T& val) {
	for (int i = 0; i < vec.size(); i++) {
		if (vec[i] == val) {
			return i;
		}
	}
	return -1;
}

int main() {
	std::ifstream read("lineup.in");
	int req_num;
	read >> req_num;
	vector<vector<int>> neighbors(COWS.size());
	for (int r = 0; r < req_num; r++) {
		string cow1;
		string cow2;

> 		string trash;

Reviewer: better name pls

Author: i mean it is trash lmao

					order.push_back(at);
					
					int a = neighbors[at][0];
					int b = neighbors[at][1];
					int temp_at = a == prev ? b : a;
					prev = at;
					at = temp_at;
				}
				// Add the final element
				added[at] = true;
				order.push_back(at);
			}
		}
	}

	std::ofstream out("lineup.out");
	for (int c : order) {
		out << COWS[c] << endl;
	}
}
```

</CPPSection>
<JavaSection>

```java
import java.io.*;
import java.util.*;

public class LineUp {
	// Assumed to be in sorted order (which it is)
	static final List<String> COWS = new ArrayList<>(
		Arrays.asList(new String[] {
			"Beatrice",
			"Belinda",
			"Bella",
			"Bessie",
			"Betsy",
			"Blue",
			"Buttercup",
			"Sue"  
		})
	);

	public static void main(String[] args) throws IOException {
		BufferedReader read = new BufferedReader(new FileReader("lineup.in"));
		int reqNum = Integer.parseInt(read.readLine());

		List<Integer>[] neighbors = new ArrayList[COWS.size()];
		for (int c = 0; c < COWS.size(); c++) {
			neighbors[c] = new ArrayList<>();
		}
		for (int r = 0; r < reqNum; r++) {
			String[] words = read.readLine().split(" ");

			// Convert the names to their index in the list
			int cow1 = COWS.indexOf(words[0]);

> 			int cow2 = COWS.indexOf(words[words.length - 1]);

Reviewer 2: this makes the solution O(n^2)

Author: <img width="429" alt="image" src="https://user-images.githubusercontent.com/55369003/154201107-fc1b4ac7-6edb-47db-9104-cb3b2f3b74d5.png">
well, benq's solution does this as well
but ok i'll uh use a hashmap

Author: fixed
